rest/utilities_testing.go

```diff
@@ -0,0 +1,245 @@
+			Origin:      []string{"http://example.com", "*", "http://staging.example.com"},
+			LoginOrigin: []string{"http://example.com"},
+			Headers:     []string{},
+			MaxAge:      1728000,
+		}
+
+		rt.RestTesterServerContext = NewServerContext(&ServerConfig{
+			CORS:           corsConfig,
+			Facebook:       &FacebookConfig{},
+			AdminInterface: &DefaultAdminInterface,
+		})
+
+		_, err := rt.RestTesterServerContext.AddDatabaseFromConfig(&DbConfig{
+			BucketConfig: BucketConfig{
+				Server:   &server,
+				Bucket:   &spec.BucketName,
+				Username: username,
+				Password: password,
+			},
+
+			Name:        "db",
+			Sync:        syncFnPtr,
+			CacheConfig: rt.CacheConfig,
+		})
+		if err != nil {
+			panic(fmt.Sprintf("Error from AddDatabaseFromConfig: %v", err))
+		}
+		rt.RestTesterBucket = rt.RestTesterServerContext.Database("db").Bucket
+
+		if !rt.noAdminParty {
+			rt.SetAdminParty(true)
+		}
+
+	}
+	return rt.RestTesterBucket
+}
+
+func (rt *RestTester) BucketAllowEmptyPassword() base.Bucket {
+
+	//Create test DB with "AllowEmptyPassword" true
+	server := "walrus:"
+	bucketName := fmt.Sprintf("sync_gateway_test_%d", gBucketCounter)
+	gBucketCounter++
+
+	rt.RestTesterServerContext = NewServerContext(&ServerConfig{
+		CORS:           &CORSConfig{},
+		Facebook:       &FacebookConfig{},
+		AdminInterface: &DefaultAdminInterface,
+	})
+
+	_, err := rt.RestTesterServerContext.AddDatabaseFromConfig(&DbConfig{
+		BucketConfig: BucketConfig{
+			Server: &server,
+			Bucket: &bucketName},
+		Name:               "db",
+		AllowEmptyPassword: true,
+		CacheConfig:        rt.CacheConfig,
+	})
+
+	if err != nil {
+		panic(fmt.Sprintf("Error from AddDatabaseFromConfig: %v", err))
+	}
+	rt.RestTesterBucket = rt.RestTesterServerContext.Database("db").Bucket
+
+	return rt.RestTesterBucket
+}
+
+func (rt *RestTester) ServerContext() *ServerContext {
+	rt.Bucket()
+	return rt.RestTesterServerContext
+}
+
+// Returns first database found for server context.
+func (rt *RestTester) GetDatabase() *db.DatabaseContext {
+
+	for _, database := range rt.ServerContext().AllDatabases() {
+		return database
+	}
+	return nil
+}
+
+func (rt *RestTester) WaitForSequence(seq uint64) error {
+	database := rt.GetDatabase()
+	if database == nil {
+		return fmt.Errorf("No database found")
+	}
+	return database.WaitForSequence(seq)
+}
+
+func (rt *RestTester) WaitForPendingChanges() error {
+	database := rt.GetDatabase()
+	if database == nil {
+		return fmt.Errorf("No database found")
+	}
+	return database.WaitForPendingChanges()
+}
+
+func (rt *RestTester) SetAdminParty(partyTime bool) {
+	a := rt.ServerContext().Database("db").Authenticator()
+	guest, _ := a.GetUser("")
+	guest.SetDisabled(!partyTime)
+	var chans channels.TimedSet
+	if partyTime {
+		chans = channels.AtSequence(base.SetOf("*"), 1)
+	}
+	guest.SetExplicitChannels(chans)
+	a.Save(guest)
+}
+
+func (rt *RestTester) Close() {
+	if rt.RestTesterServerContext != nil {
+		rt.RestTesterServerContext.Close()
```

Comment:
Maybe worth noting regarding the test interference - I think that ServerContext.Close() includes some async shutdown processing (like the feed runner goroutine) that might be a problem.